File: backend/schemas/market.py
"""
Pydantic schemas for Marketplace API
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field


# ================== MarketItem Schemas ==================

class MarketItemCreate(BaseModel):
    """Схема для создания товара на маркетплейсе"""
    item_type: str = Field(..., description="Тип: 'template' или 'scenario'")
    source_bot_id: Optional[int] = Field(None, description="ID бота-источника (для шаблонов)")
    source_scenario_id: Optional[int] = Field(None, description="ID сценария-источника")
    title: str = Field(..., max_length=255, description="Название товара")
    description: Optional[str] = Field(None, description="Основное описание")
    additional_description: Optional[str] = Field(None, description="Дополнительное описание")
    image_url: Optional[str] = Field(None, max_length=500, description="URL изображения")
    price: Decimal = Field(0.00, ge=0, description="Цена (0 = бесплатно)")
    category: Optional[str] = Field(None, max_length=100, description="Категория")
    tags: Optional[List[str]] = Field(None, description="Список тегов")
    is_premium: bool = Field(False, description="Premium товар")
    is_published: bool = Field(False, description="Опубликован ли товар")


class MarketItemUpdate(BaseModel):
    """Схема для обновления товара"""
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    additional_description: Optional[str] = None
    image_url: Optional[str] = Field(None, max_length=500)
    price: Optional[Decimal] = Field(None, ge=0)
    category: Optional[str] = Field(None, max_length=100)
    tags: Optional[List[str]] = None
    is_premium: Optional[bool] = None
    is_active: Optional[bool] = None
    is_published: Optional[bool] = None


class SellerInfo(BaseModel):
    """Информация о продавце"""
    id: int
    name: Optional[str] = None
    email: str
    avatar: Optional[str] = None
    
    class Config:
        from_attributes = True


class MarketItemOut(BaseModel):
    """Схема для вывода товара"""
    id: int
    item_type: str
    source_bot_id: Optional[int] = None
    source_scenario_id: Optional[int] = None
    title: str
    description: Optional[str] = None
    additional_description: Optional[str] = None
    image_url: Optional[str] = None
    price: Decimal
    sales_count: int
    category: Optional[str] = None
    tags: Optional[List[str]] = None
    is_premium: bool
    is_active: bool
    is_published: bool
    seller: SellerInfo
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime] = None
    # Вычисляемые поля
    average_rating: Optional[float] = None
    rating_count: int = 0
    
    class Config:
        from_attributes = True


class MarketItemDetailOut(MarketItemOut):
    """Детальная информация о товаре"""
    reviews: List['MarketReviewOut'] = []


class MyTemplateOut(BaseModel):
    """Шаблон разработчика для кабинета (GET /my-templates)"""
    id: int
    name: str
    status: str  # published | draft (legacy, для совместимости)
    moderation_status: str  # draft | pending | approved | rejected
    moderation_rejection_reason: Optional[str] = None
    installs_count: int
    views_count: int
    created_at: datetime


# ================== MarketOrder Schemas ==================

class MarketOrderCreate(BaseModel):
    """Схема для создания заказа"""
    title: str = Field(..., max_length=255, description="Название заказа")
    description: str = Field(..., description="Описание требований")
    budget_min: Optional[Decimal] = Field(None, ge=0, description="Минимальный бюджет")
    budget_max: Optional[Decimal] = Field(None, ge=0, description="Максимальный бюджет")
    deadline: Optional[datetime] = Field(None, description="Срок выполнения")
    category: Optional[str] = Field(None, max_length=100, description="Категория")
    skills: Optional[List[str]] = Field(None, description="Требуемые навыки")


class MarketOrderUpdate(BaseModel):
    """Схема для обновления заказа"""
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    budget_min: Optional[Decimal] = Field(None, ge=0)
    budget_max: Optional[Decimal] = Field(None, ge=0)
    deadline: Optional[datetime] = None
    category: Optional[str] = Field(None, max_length=100)
    skills: Optional[List[str]] = None
    status: Optional[str] = None  # open, in_progress, completed, cancelled


class MarketOrderOut(BaseModel):
    """Схема для вывода заказа"""
    id: int
    title: str
    description: str
    budget_min: Optional[Decimal] = None
    budget_max: Optional[Decimal] = None
    deadline: Optional[datetime] = None
    category: Optional[str] = None
    skills: Optional[List[str]] = None
    status: str
    author: SellerInfo
    selected_freelancer: Optional[SellerInfo] = None
    created_at: datetime
    updated_at: datetime
    proposals_count: int = 0
    
    class Config:
        from_attributes = True


class MarketOrderDetailOut(MarketOrderOut):
    """Детальная информация о заказе"""
    proposals: List['OrderProposalOut'] = []


# ================== OrderProposal Schemas ==================

class OrderProposalCreate(BaseModel):
    """Схема для создания предложения на заказ"""
    message: Optional[str] = Field(None, description="Сообщение от исполнителя")
    proposed_price: Optional[Decimal] = Field(None, ge=0, description="Предлагаемая цена")
    estimated_days: Optional[int] = Field(None, ge=1, description="Оценочное количество дней")


class OrderProposalOut(BaseModel):
    """Схема для вывода предложения"""
    id: int
    order_id: int
    freelancer: SellerInfo
    message: Optional[str] = None
    proposed_price: Optional[Decimal] = None
    estimated_days: Optional[int] = None
    is_accepted: bool
    is_declined: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ================== FreelancerProfile Schemas ==================

class FreelancerProfileCreate(BaseModel):
    """Схема для создания профиля исполнителя"""
    title: str = Field(..., max_length=255, description="Название специализации")
    description: Optional[str] = Field(None, description="Описание услуг")
    hourly_rate: Optional[Decimal] = Field(None, ge=0, description="Стоимость за час")
    skills: Optional[List[str]] = Field(None, description="Список навыков")
    portfolio_items: Optional[List[str]] = Field(None, description="Примеры работ")


class FreelancerProfileUpdate(BaseModel):
    """Схема для обновления профиля исполнителя"""
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    hourly_rate: Optional[Decimal] = Field(None, ge=0)
    skills: Optional[List[str]] = None
    portfolio_items: Optional[List[str]] = None
    is_active: Optional[bool] = None


class FreelancerProfileOut(BaseModel):
    """Схема для вывода профиля исполнителя"""
    id: int
    user: SellerInfo
    title: str
    description: Optional[str] = None
    hourly_rate: Optional[Decimal] = None
    skills: Optional[List[str]] = None
    portfolio_items: Optional[List[str]] = None
    completed_orders_count: int
    is_active: bool
    is_verified: bool
    created_at: datetime
    updated_at: datetime
    # Вычисляемые поля
    average_rating: Optional[float] = None
    rating_count: int = 0
    
    class Config:
        from_attributes = True


# ================== MarketReview Schemas ==================

class MarketReviewCreate(BaseModel):
    """Схема для создания отзыва"""
    item_type: str = Field(..., description="Тип: 'market_item', 'market_order', 'freelancer'")
    item_id: int = Field(..., description="ID товара/заказа/исполнителя")
    rating: int = Field(..., ge=1, le=5, description="Рейтинг от 1 до 5")
    comment: Optional[str] = Field(None, description="Текст отзыва")


class MarketReviewOut(BaseModel):
    """Схема для вывода отзыва"""
    id: int
    item_type: str
    item_id: int
    author: SellerInfo
    rating: int
    comment: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ================== Response Schemas ==================

class MarketItemListResponse(BaseModel):
    """Ответ со списком товаров"""
    total: int
    items: List[MarketItemOut]
    page: int = 1
    page_size: int = 20


class MarketOrderListResponse(BaseModel):
    """Ответ со списком заказов"""
    total: int
    items: List[MarketOrderOut]
    page: int = 1
    page_size: int = 20


class FreelancerListResponse(BaseModel):
    """Ответ со списком исполнителей"""
    total: int
    items: List[FreelancerProfileOut]
    page: int = 1
    page_size: int = 20


# Обновляем forward references после определения всех классов
MarketItemDetailOut.model_rebuild()
MarketOrderDetailOut.model_rebuild()

File: frontend/src/api/market.ts
/**
 * API клиент для маркетплейса
 */
import api from './client';

// ================== Types ==================

export interface MarketItem {
  id: number;
  item_type: 'template' | 'scenario';
  source_bot_id?: number;
  source_scenario_id?: number;
  title: string;
  description?: string;
  additional_description?: string;
  image_url?: string;
  price: number;
  sales_count: number;
  category?: string;
  tags?: string[];
  is_premium: boolean;
  is_active: boolean;
  is_published: boolean;
  seller: {
    id: number;
    name?: string;
    email: string;
    avatar?: string;
  };
  created_at: string;
  updated_at: string;
  published_at?: string;
  average_rating?: number;
  rating_count: number;
}

export interface MarketItemCreate {
  item_type: 'template' | 'scenario';
  source_bot_id?: number;
  source_scenario_id?: number;
  title: string;
  description?: string;
  additional_description?: string;
  image_url?: string;
  price: number;
  category?: string;
  tags?: string[];
  is_premium?: boolean;
  is_published?: boolean;
}

export interface MarketItemListResponse {
  total: number;
  items: MarketItem[];
  page: number;
  page_size: number;
}

export interface MyTemplate {
  id: number;
  name: string;
  status: 'published' | 'draft';
  moderation_status: 'draft' | 'pending' | 'approved' | 'rejected';
  moderation_rejection_reason?: string | null;
  installs_count: number;
  views_count: number;
  created_at: string;
}

export interface MarketOrder {
  id: number;
  title: string;
  description: string;
  budget_min?: number;
  budget_max?: number;
  deadline?: string;
  category?: string;
  skills?: string[];
  status: 'open' | 'in_progress' | 'completed' | 'cancelled';
  author: {
    id: number;
    name?: string;
    email: string;
    avatar?: string;
  };
  selected_freelancer?: {
    id: number;
    name?: string;
    email: string;
    avatar?: string;
  };
  created_at: string;
  updated_at: string;
  proposals_count: number;
}

export interface MarketOrderCreate {
  title: string;
  description: string;
  budget_min?: number;
  budget_max?: number;
  deadline?: string;
  category?: string;
  skills?: string[];
}

export interface MarketOrderListResponse {
  total: number;
  items: MarketOrder[];
  page: number;
  page_size: number;
}

export interface FreelancerProfile {
  id: number;
  user: {
    id: number;
    name?: string;
    email: string;
    avatar?: string;
  };
  title: string;
  description?: string;
  hourly_rate?: number;
  skills?: string[];
  portfolio_items?: string[];
  completed_orders_count: number;
  is_active: boolean;
  is_verified: boolean;
  created_at: string;
  updated_at: string;
  average_rating?: number;
  rating_count: number;
}

export interface FreelancerProfileCreate {
  title: string;
  description?: string;
  hourly_rate?: number;
  skills?: string[];
  portfolio_items?: string[];
}

export interface MarketReview {
  id: number;
  item_type: string;
  item_id: number;
  author: {
    id: number;
    name?: string;
    email: string;
    avatar?: string;
  };
  rating: number;
  comment?: string;
  created_at: string;
  updated_at: string;
}

export interface MarketReviewCreate {
  item_type: 'market_item' | 'market_order' | 'freelancer';
  item_id: number;
  rating: number;
  comment?: string;
}

// ================== Creator Dashboard (Developer plan) ==================

/**
 * Получить список своих шаблонов (кабинет разработчика).
 * Требует тариф Developer.
 */
export async function getMyTemplates(): Promise<MyTemplate[]> {
  return api.get('/api/market/my-templates');
}

/**
 * Отправить шаблон на модерацию (draft → pending).
 * Требует тариф Developer.
 */
export async function submitTemplateToModeration(
  templateId: number
): Promise<{ ok: boolean; moderation_status: string }> {
  return api.post(`/api/market/templates/${templateId}/submit`);
}

// ================== Market Items ==================

/**
 * Получить список товаров на маркетплейсе
 */
export async function getMarketItems(params?: {
  item_type?: 'template' | 'scenario';
  category?: string;
  search?: string;
  min_price?: number;
  max_price?: number;
  is_premium?: boolean;
  is_published?: boolean;
  sort_by?: string;
  order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}): Promise<MarketItemListResponse> {
  try {
    const searchParams = new URLSearchParams();
    if (params) {
      if (params.item_type) searchParams.append('item_type', params.item_type);
      if (params.category) searchParams.append('category', params.category);
      if (params.search) searchParams.append('search', params.search);
      if (params.min_price !== undefined)
        searchParams.append('min_price', params.min_price.toString());
      if (params.max_price !== undefined)
        searchParams.append('max_price', params.max_price.toString());
      if (params.is_premium !== undefined)
        searchParams.append('is_premium', params.is_premium.toString());
      if (params.is_published !== undefined)
        searchParams.append('is_published', params.is_published.toString());
      if (params.sort_by) searchParams.append('sort_by', params.sort_by);
      if (params.order) searchParams.append('order', params.order);
      if (params.page !== undefined) searchParams.append('page', params.page.toString());
      if (params.page_size !== undefined)
        searchParams.append('page_size', params.page_size.toString());
    }
    const url = `/api/market/items${searchParams.toString() ? `?${searchParams.toString()}` : ''}`;
    return await api.get(url);
  } catch (error: any) {
    console.error('Failed to fetch market items:', error);
    throw error;
  }
}

/**
 * Получить детальную информацию о товаре
 */
export async function getMarketItem(itemId: number): Promise<MarketItem> {
  try {
    return await api.get(`/api/market/items/${itemId}`);
  } catch (error: any) {
    console.error('Failed to fetch market item:', error);
    throw error;
  }
}

/**
 * Создать товар на маркетплейсе
 */
export async function createMarketItem(data: MarketItemCreate): Promise<MarketItem> {
  try {
    return await api.post('/api/market/items', data);
  } catch (error: any) {
    console.error('Failed to create market item:', error);
    throw error;
  }
}

/**
 * Обновить товар на маркетплейсе
 */
export async function updateMarketItem(
  itemId: number,
  data: Partial<MarketItemCreate>
): Promise<MarketItem> {
  try {
    return await api.put(`/api/market/items/${itemId}`, data);
  } catch (error: any) {
    console.error('Failed to update market item:', error);
    throw error;
  }
}

/**
 * Удалить товар с маркетплейса
 */
export async function deleteMarketItem(itemId: number): Promise<void> {
  try {
    await api.delete(`/api/market/items/${itemId}`);
  } catch (error: any) {
    console.error('Failed to delete market item:', error);
    throw error;
  }
}

// ================== Market Orders ==================

/**
 * Получить список заказов
 */
export async function getMarketOrders(params?: {
  status?: string;
  category?: string;
  search?: string;
  sort_by?: string;
  order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}): Promise<MarketOrderListResponse> {
  try {
    return await api.get('/api/market/orders', params);
  } catch (error: any) {
    console.error('Failed to fetch market orders:', error);
    throw error;
  }
}

/**
 * Создать заказ
 */
export async function createMarketOrder(data: MarketOrderCreate): Promise<MarketOrder> {
  try {
    return await api.post('/api/market/orders', data);
  } catch (error: any) {
    console.error('Failed to create market order:', error);
    throw error;
  }
}

// ================== Freelancer Profiles ==================

/**
 * Получить список исполнителей
 */
export async function getFreelancers(params?: {
  search?: string;
  skills?: string;
  is_verified?: boolean;
  sort_by?: string;
  order?: 'asc' | 'desc';
  page?: number;
  page_size?: number;
}): Promise<{ total: number; items: FreelancerProfile[]; page: number; page_size: number }> {
  try {
    return await api.get('/api/market/freelancers', params);
  } catch (error: any) {
    console.error('Failed to fetch freelancers:', error);
    throw error;
  }
}

/**
 * Создать профиль исполнителя
 */
export async function createFreelancerProfile(
  data: FreelancerProfileCreate
): Promise<FreelancerProfile> {
  try {
    return await api.post('/api/market/freelancers', data);
  } catch (error: any) {
    console.error('Failed to create freelancer profile:', error);
    throw error;
  }
}

// ================== Reviews ==================

/**
 * Создать отзыв
 */
export async function createMarketReview(data: MarketReviewCreate): Promise<MarketReview> {
  try {
    return await api.post('/api/market/reviews', data);
  } catch (error: any) {
    console.error('Failed to create review:', error);
    throw error;
  }
}

/**
 * Получить список отзывов
 */
export async function getMarketReviews(
  itemType: string,
  itemId: number,
  params?: { page?: number; page_size?: number }
): Promise<MarketReview[]> {
  try {
    return await api.get('/api/market/reviews', {
      item_type: itemType,
      item_id: itemId,
      ...params,
    });
  } catch (error: any) {
    console.error('Failed to fetch reviews:', error);
    throw error;
  }
}
